### PR TITLE
collapse admission chains

### PR DESCRIPTION
**Pull request by deads2k**
This collapses the separate origin and kube admission chains into a single chain that is used twice.  There is a lot of code around figuring out whether you're allowed to collapse the chains, because this is not always allowed.

I plan to update configapi validation to warn if you have anything that prevents your chains from being combined.  Eventually I'd like to make it error, but we need to complete the default on/off combinations first.

@csrwng yours I think

### CODE REVIEW

**Comment by deads2k** [comment]
[test]

File: pkg/cmd/server/origin/master_config.go [hunk 1]
```diff
@@ -309,6 +285,200 @@ func BuildMasterConfig(options configapi.MasterConfig) (*MasterConfig, error) {
...
+}
+
+// newAdmissionChainFunc is for unit testing only.  You should NEVER OVERRIDE THIS outside of a unit test.
+var newAdmissionChainFunc = newAdmissionChain
+
+func newAdmissionChain(pluginNames []string, admissionConfigFilename string, pluginConfig map[string]configapi.AdmissionPluginConfig, options configapi.MasterConfig, kubeClientSet *internalclientset.Clientset, pluginInitializer oadmission.PluginInitializer) (admission.Interface, error) {
+	plugins := []admission.Interface{}
+	for _, pluginName := range pluginNames {
+		switch pluginName {
+		case lifecycle.PluginName:
```
Comment (csrwng) [inline pkg/cmd/server/origin/master_config.go]: Maybe outside of the scope of this pull, but shouldn't these cases be handled by WantsXXX type interfaces?

File: pkg/cmd/server/origin/master_config.go [hunk 2]
```diff
@@ -309,6 +285,200 @@ func BuildMasterConfig(options configapi.MasterConfig) (*MasterConfig, error) {
...
+		"SecurityContextConstraint",
+		"BuildDefaults",
+		"BuildOverrides",
+		"AlwaysPullImages",
+		"LimitPodHardAntiAffinityTopology",
+		"SCCExecRestrictions",
+		"ResourceQuota",
+	}
+
+	combinedAdmissionControlPlugins = []string{
```
Comment (csrwng) [inline pkg/cmd/server/origin/master_config.go]: Besides the unit test to validate these, a comment would be helpful to let anyone that's editing these chains know that they need to be kept in sync.

File: pkg/cmd/server/origin/master_config.go (hunk 1 above)
Comment (deads2k) [inline pkg/cmd/server/origin/master_config.go]: > Maybe outside of the scope of this pull, but shouldn't these cases be handled by WantsXXX type interfaces?

Outside the scope.  This looks like a good case, but what object would we expose? The entire config seems prone to mis-use.

**Comment by csrwng** [comment]
Other than the couple of comments above, LGTM

File: pkg/cmd/server/origin/master_config.go (hunk 1 above)
Comment (csrwng) [inline pkg/cmd/server/origin/master_config.go]: I would assume we pass exactly the data they need (immortal namespaces, reject/admit CIDRs, LimitSecretReferences, etc). This is what we do with the other plugins, no?

Comment (deads2k) [inline pkg/cmd/server/origin/master_config.go]: > I would assume we pass exactly the data they need (immortal namespaces, reject/admit CIDRs, LimitSecretReferences, etc). This is what we do with the other plugins, no?

We've tried to keep the interfaces generally useful (think passing client.Interface, not individual namespacers).  We could try to build a synthetic `AdmissionPluginConfig` for them, which would actually fit quite well.

**Comment by deads2k** [comment]
comments addressed, opened https://github.com/openshift/origin/issues/9738, holding on rebase

**Comment by deads2k** [comment]
@csrwng I had to make the upstream namespace lifecycle namespace aware to avoid leaking information about namespaces during access review checks.  Kube won't experience the issue until a couple pulls after https://github.com/kubernetes/kubernetes/pull/20573 when they get namespaced access review checks.  

Double check to make sure you're ok with the upstream carry/todo.

**Comment by csrwng** [comment]
@deads2k the upstream carry LGTM, however, it's a little worrying that we can be broken by any upstream change or new plugin that assumes that only upstream objects are going to be sent to it. Not sure what's the best way to mitigate that risk though.

**Comment by deads2k** [comment]
> @deads2k the upstream carry LGTM, however, it's a little worrying that we can be broken by any upstream change or new plugin that assumes that only upstream objects are going to be sent to it. Not sure what's the best way to mitigate that risk though.

We should update the admission interface to require `admission.Handle` and improve `Handle` to return back a list of resources or a filter func.  Any filter func should be flagged via a reflective test in origin to force inspection.  @derekwaynecarr disagree?

**Comment by derekwaynecarr** [comment]
@deads2k - seems reasonable, what would you want `admission.Handles` signature to look like?  Do you want filtering per `Operation`?

**Comment by deads2k** [comment]
> @deads2k - seems reasonable, what would you want admission.Handles signature to look like? Do you want filtering per Operation?

```
type Handler interface{
     func Operations() sets.String
     // Resource returns the set of resources this admission plugin may be interested in.
     // * in group means any group, * in resource means any resource
     func Resources() map[GroupResource]bool
     // Filter is called if the Operations and Resources match to further restrict the set of allowed
     // admission attributes
     func Filter() func(admission.Attributes) bool
}

```


I think that covers it.

**Comment by derekwaynecarr** [comment]
So basically, the flow is:

1. do you care about this group resource
2. do you care about this operation
3. do you care about this particular request (fitler piece)
4. if all 3 above hold true, then call Admit

that seems reasonable at first glance to me.

I am interested in breaking out `Admit` to also say:

1. do you need to modify this object (call ApplyDefaults)
2. do you need to enforce constraints (call ApplyConstraints)
3. do you need to admit/reject (call Admit)

It would be good if we add filtering that we do it at the same time as also breaking out the other patterns around mutation, constraint enforcement, and then admit/reject (ideally post object validation).

I am going to try and advocate for a card in the 1.4 work to push this.

**Comment by deads2k** [comment]
superceded

**Comment by smarterclayton** [comment]
What am I supposed to do with this carry in 1.4?

**Comment by smarterclayton** [comment]
Do I take https://github.com/kubernetes/kubernetes/pull/29926 instead?

**Comment by deads2k** [comment]
> What am I supposed to do with this carry in 1.4?

Carry it.  I want to fix it in discovery and I haven't refactored it upstream yet.

**Pull request closed without merging**